src/code/cutscene_camera.c

```diff
@@ -353,28 +357,579 @@ void CutsceneCamera_Reset(void) {
+
+        *camRoll += (s16)(temp * lerp);
+    }
+
+    if (interpState->curFrame >= pointCmd->duration) {
+        // Finished
+        interpState->type = CS_CAM_INTERP_OFF;
+        return 1;
+    }
+
+    return 0;
+}
+
+s16 CutsceneCamera_Interp_Geo(Vec3f* camPos, f32* camFov, s16* camRoll, CsCmdCamPoint* pointCmd, CsCmdCamMisc* miscCmd,
+                              CutsceneCameraInterp* interpState) {
+    VecGeo sp40;
+    f32 sp3C;
+    f32 tmp1;
+    f32 tmp2;
+
+    if (interpState->type != CS_CAM_INTERP_GEO) {
+        // Initialize
+        interpState->type = CS_CAM_INTERP_GEO;
+        interpState->waypoint = 0;
+        interpState->curFrame = 0;
+        interpState->duration = 1;
+        if (camPos != NULL) {
+            interpState->unk_20 = OLib_Vec3fDist(&interpState->curPos, camPos) * Math_FTanF(DEG_TO_RAD(*camFov));
+        }
+        if (camFov != NULL) {
+            interpState->initFov = *camFov;
+        }
+        if (camRoll != NULL) {
+            interpState->initRoll = *camRoll;
+        }
+    }
+
+    tmp2 = (((pointCmd->weight + 100) * (pointCmd->duration / 2)) +
+            (((pointCmd->weight + 100) / 2) * (pointCmd->duration & 1)));
+    if (pointCmd->duration < 2) {
+        sp3C = 1.0f;
+    } else {
+        tmp1 = (f32)(pointCmd->weight - 100) / (pointCmd->duration - 1);
+        sp3C = ((interpState->curFrame * tmp1) + 100.0f) / tmp2;
+    }
+
+    interpState->curFrame++;
+
+    if (camPos != NULL) {
+        OLib_Vec3fDiffToVecGeo(&sp40, &interpState->curPos, camPos);
+        sp40.r = interpState->unk_20 / Math_FTanF(DEG_TO_RAD(*camFov));
+        OLib_AddVecGeoToVec3f(camPos, &interpState->curPos, &sp40);
+    }
+
+    if (camFov != NULL) {
+        *camFov += (miscCmd->fov - interpState->initFov) * sp3C;
+    }
+
+    if (camRoll != NULL) {
+        s16 new_var;
+        s16 temp;
+
+        new_var = CAM_DEG_TO_BINANG(miscCmd->roll);
+
+        temp = (s16)(new_var - (s16)interpState->initRoll);
+
+        *camRoll += (s16)(temp * sp3C);
+    }
+
+    if (interpState->curFrame >= pointCmd->duration) {
+        // Finished
+        interpState->type = CS_CAM_INTERP_OFF;
+        return 1;
+    }
+
+    return 0;
+}
+
+// Updates the interpolation state but does not change the pos/fov/roll values
+s16 CutsceneCamera_Interp_None(Vec3f* camPos, f32* camFov, s16* camRoll, CsCmdCamPoint* pointCmd, CsCmdCamMisc* miscCmd,
+                               CutsceneCameraInterp* interpState) {
+    if (interpState->type != CS_CAM_INTERP_NONE) {
+        // Initialize
+        interpState->type = CS_CAM_INTERP_NONE;
+        interpState->waypoint = 0;
+        interpState->curFrame = 0;
+        interpState->duration = 1;
+    }
+
+    interpState->curFrame++;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/cutscene_camera/func_80161E4C.s")
+    if (interpState->curFrame >= pointCmd->duration) {
+        // Finishes
+        interpState->type = CS_CAM_INTERP_OFF;
+        return true;
```

Comment:
Hmm, the other Interp functions seem to suggest these aren't necessarily just true/false, but duration?

Reply:
Ah, yes I thought I caught all these. You're right, it's not a true/false return